### DISCUSSION TITLE
R is aborted when fitting a 1-compartment model

### DISCUSSION THREAD

**Discussion by dryoutao** [nlmixrdevelopment/nlmixr#569] [Q&A]
Hi,

This is a Mac user. I changed to a new MacBook Pro with R v4.0.5 (2021-03-31) -- "Shake and Throw".

RxODE v1.1.1 was installed.

nlmixr v2.0.5 was installed by running install.packages("nlmixr"). I did not install the development version.

When I run the following:

## Load libraries
library(ggplot2)
library(RxODE)
library(nlmixr)

str(theo_sd)
ggplot(theo_sd, aes(TIME, DV)) +
  geom_line(aes(group=ID), col="red") +
  scale_x_continuous("Time (h)") +
  scale_y_continuous("Concentration") +
  labs(title="Theophylline single-dose", subtitle="Concentration vs. time by individual")

one.cmt <- function() {
  ini({
    tka <- .5   # log Ka
    tcl <- -3.2 # log Cl
    tv <- -1    # log V
    eta.ka ~ 1
    eta.cl ~ 2
    eta.v ~ 1
    add.err <- 0.1
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    linCmt() ~ add(add.err)
  })
}
fit <- nlmixr(one.cmt, theo_sd, est="nlme")

I get an error message

 parameter labels from comments will be replaced by 'label()'
Error in MEEM(object, conLin, control$niterEM) : 
  Singularity in backsolve at level 0, block 1

When I try this: 

fit <- nlmixr(one.cmt, theo_sd, est="foce")

R is aborted.

What do I need to do? I'd be very grateful!

Tao

**Comment by mattfidler** [comment]
Hi @dryoutao,

This could be `symengine`; try:

```r
library(symengine)
```

If this crashes R, then this is the issue.

The work-around is described below


https://github.com/nlmixrdevelopment/RxODE#mac-osx

**Reply by dryoutao** [reply]
Many thanks, Matt, for the prompt reply! Really appreciate this!

Indeed running "library(symengine)" crashed R. I have followed the work-around and library(symengine) was successfully loaded.

However, when I run these:

## Load libraries
library(ggplot2)
library(symengine)
library(RxODE)
library(nlmixr)

str(theo_sd)
ggplot(theo_sd, aes(TIME, DV)) +
  geom_line(aes(group=ID), col="red") +
  scale_x_continuous("Time (h)") +
  scale_y_continuous("Concentration") +
  labs(title="Theophylline single-dose", subtitle="Concentration vs. time by individual")

one.cmt <- function() {
  ini({
    tka <- .5   # log Ka
    tcl <- -3.2 # log Cl
    tv <- -1    # log V
    eta.ka ~ 1
    eta.cl ~ 2
    eta.v ~ 1
    add.err <- 0.1
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    linCmt() ~ add(add.err)
  })
}

########################################
fit <- nlmixr(one.cmt, theo_sd, est="nlme")

I got 

ℹ parameter labels from comments will be replaced by 'label()'
Error in MEEM(object, conLin, control$niterEM) : 
  Singularity in backsolve at level 0, block 1

########################################
when running these:
> fit <- nlmixr(one.cmt, theo_sd, est="nlme", calc.resid=FALSE)

I got 

ℹ parameter labels from comments will be replaced by 'label()'
Error in (function (model, data = sys.frame(sys.parent()), fixed, random = fixed,  : 
  unused argument (calc.resid = FALSE)

########################################
when running these:

fit <- nlmixr(one.cmt, theo_sd, est="foce")

I got

ℹ parameter labels from comments will be replaced by 'label()'
→ creating full model...
→ pruning branches (`if`/`else`)...
✓ done
→ loading into symengine environment...
✓ done
→ creating full model...
→ pruning branches (`if`/`else`)...
✓ done
→ loading into symengine environment...
✓ done
→ calculate jacobian

→ calculate ∂(f)/∂(η)
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

→ finding duplicate expressions in inner model...
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

→ optimizing duplicate expressions in inner model...
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

→ finding duplicate expressions in EBE model...
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

→ optimizing duplicate expressions in EBE model...
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

→ compiling inner model...
✓ done
→ finding duplicate expressions in FD model...
[====|====|====|====|====|====|====|====|====|====] 0:00:00 

→ compiling EBE model...
✓ done
→ compiling events FD model...
✓ done
Key: U: Unscaled Parameters; X: Back-transformed parameters; G: Gill difference gradient approximation
F: Forward difference gradient approximation
C: Central difference gradient approximation
M: Mixed forward and central difference gradient approximation
Unscaled parameters for Omegas=chol(solve(omega));
Diagonals are transformed, as specified by foceiControl(diagXform=)
|-----+---------------+-----------+-----------+-----------+-----------+-----------+-----------+-----------|
|    #| Objective Fun |       tka |       tcl |        tv |   add.err |        o1 |        o2 |        o3 |
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
|    1|     49559.473 |     1.000 |    -1.000 |   -0.6769 |   -0.5099 |   -0.5072 |   -0.5077 |   -0.5072 |
|    U|     49559.473 |     437.7 |    -141.8 |    -48.19 |    0.1981 |     1.000 |    0.8409 |     1.000 |
|    X|     49559.473 |1.244e+190 | 2.612e-62 | 1.183e-21 |    0.1981 |     1.000 |    0.8409 |     1.000 |
|    G|    Gill Diff. | -0.001953 |     0.000 |    -1254. |    -8579. | -0.002592 |     0.000 | 1.310e+05 |
Error in .fitFun(.ret) : 
  Evaluation error: On initial gradient evaluation, one or more parameters have a zero gradient
Change model, try different initial estimates or use outerOpt="bobyqa").
Restart 1
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
Theta reset (ETA drift)
|    1|     73859.995 |     1.000 |    -1.000 |   -0.7313 |   -0.5967 |   -0.5934 |   -0.5940 |   -0.5934 |
|    U|     73859.995 |     434.9 |    -109.7 |    -36.54 |    0.1110 |     1.000 |    0.8409 |     1.000 |
|    X|     73859.995 |7.154e+188 | 2.287e-48 | 1.350e-16 |    0.1110 |     1.000 |    0.8409 |     1.000 |
|    G|    Gill Diff. |     0.000 |     0.000 |    -974.6 |-4.902e+04 |  0.004903 | -0.004901 | 7.912e+04 |
Error in .fitFun(.ret) : 
  Evaluation error: On initial gradient evaluation, one or more parameters have a zero gradient
Change model, try different initial estimates or use outerOpt="bobyqa").
Restart 2
Error in .fitFun(.ret) : 
  Not compatible with requested type: [type=list; target=double].
Restart 3
Error in .fitFun(.ret) : 
  Not compatible with requested type: [type=list; target=double].
Error in (function (data, inits, PKpars, model = NULL, pred = NULL, err = NULL,  : 
  Could not fit data.

What should I do?

Many thanks,
Tao

**Reply by mattfidler** [reply]
This is often caused by gcc 8 on certain systems.  It occurs because of a mismatch between how Fortran and C send functions between each other.  

You can see if recompiling the whole package from source could help.  That way the fortran and C interaction are possibly consistent.  Otherwise, I would suggest using gcc and gfortran 10 which is what CRAN tests against

**Reply by mattfidler** [reply]
Hopefully this was fixed for you @dryoutao

**Reply by dryoutao** [reply]
Dear Matt,

Many thanks for the suggestion. I have downgraded GCC to v10 using brew

1. Run the following: "brew install gcc@10"

2. Check the version: "gcc --version"

Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: x86_64-apple-darwin20.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

However, the problem persists (i.e. no change at all). Is there anything else I should do?

Much appreciated!
Tao

**Reply by mattfidler** [reply]
Hm.  Did you change your Makevars?  Sometimes that helps.
